Add nuget command to dotnet cli

drewgil opened the pull request:
This command (as indicated by its tests) is for passing through to the nuget dll unmodified args for
- dotnet nuget locals
- dotnet nuget push
- dotnet nuget delete
(not passing the dotnet token, of course).

I've also resolved a namespace conflict by indicating that all NuGet code under pack is NuGet.Legacy.

@eerhardt @emgarten

dnfclas commented:
Hi __@drewgil__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

eerhardt commented on src/dotnet/commands/dotnet-nuget/NuGetCommand.cs:
```diff
@@ -0,0 +1,85 @@
+{
+    public class NuGetCommand
+    {
+        private static readonly string DefaultRid = RuntimeEnvironmentRidExtensions.GetLegacyRestoreRuntimeIdentifier();
```
This doesn't appear to be used.

eerhardt commented on src/dotnet/commands/dotnet-nuget/NuGetCommand.cs:
```diff
@@ -0,0 +1,85 @@
+            return commandRunner.Run(commandArgs);
+        }
+
+        public class NuGetCommandRunner : INuGetCommandRunner
```
Does this class need to be public?  I think it can be private.

eerhardt commented on src/dotnet/commands/dotnet-nuget/NuGetCommand.cs:
```diff
@@ -0,0 +1,85 @@
+                !commandArgs[0].Equals("push", StringComparison.OrdinalIgnoreCase) &&
+                !commandArgs[0].Equals("delete", StringComparison.OrdinalIgnoreCase))
+            {
+                return 1; // unsupported argument
```
Are users going to get a help message? 

I think there are 2 approaches that can be taken here:

1. you should be adding your valid commands to the CommandLineApplication.  That way users get a decent help experience.
2. you should just be passing through the arguments unverified to NuGet, and let NuGet do the correct handling (validate args, give help experience, etc)

livarcocc replied:
If you do it like this, if NuGet adds a new command that we want to support, we will have to change this code. We should just let nuget handle that and be a pass through.

drewgil replied:
Good catch--thanks. Here is my plan--I'll go with option 2, so that we can have a "single point of truth" for the usage text, NuGet.CommandLine.XPlat.dll. We also have a Windows exe that invokes the command in this dll, so this would avoid any duplication. This way I also won't need to do any modifying of options before passing them to the NuGet dll, as <command> <option> is its argument format too. For example "dotnet nuget push --help" would be passed to the NuGet dll as "push --help", and we can light up usage text appropriately from there.

drewgil replied:
Also, @livarcocc --I want to incorporate this thought and remove the vetting as you say. When I have this change in I'll ping the PR.

eerhardt commented on src/dotnet/commands/dotnet-nuget/NuGetCommand.cs:
```diff
@@ -0,0 +1,85 @@
+            {
+                Name = "dotnet nuget",
+                FullName = ".NET nuget command passthrough",
+                Description = "Passes selected commands to the NuGet layer"
```
I'm not sure this is the best description to explain to a user what this command does.  I don't 'layer' makes sense to a user.  And what "selected commands" mean.

drewgil replied:
Agreed. I've modified the text to make it more helpful.

Sridhar-MS commented on src/dotnet/commands/dotnet-nuget/NuGetCommand.cs:
```diff
@@ -0,0 +1,83 @@
+        {
+            public int Run(string[] commandArgs)
+            {
+                var nugetAsm = typeof(NugetProgram).GetTypeInfo().Assembly;
```
We are doing something very similar here in the `restore` command. https://github.com/dotnet/cli/blob/rel/1.0.0/src/dotnet/commands/dotnet-restore/NuGet3.cs

Should we consolidate `restore` with the rest of the nuget commands? But it will break a lot of users (using `dotnet restore` instead of `dotnet nuget restore`). Let us at least try to consolidate at source code level and avoid duplicating code.

drewgil replied:
@rohit21agrawal is working on updating the pack command. I believe the plan at this point is to keep it as dotnet pack and not dotnet nuget pack, although the cli help on this point is misleading. I agree with the idea of making it consistent, but we'd like to avoid breaking everybody. If the two operate as dumb passthroughs to the nuget dll, perhaps code could be shared, and both dotnet pack and dotnet nuget pack would light up. Unsure. @harikmenon may have thoughts on this too.

drewgil replied:
This was a loose tie-in to dotnet restore :) but my point was that we're shooting for dotnet pack and dotnet restore as these are core scenarios (and already supported as dotnet [command]), and these others (delete, locals, push) are less dotnet-ish and more nuget-ish, hence the distinction in the command line.

gulbanana replied:
pack isn't very useful without push, though..

drewgil commented:
Have incorporated (and squashed) this feedback into the commit. Please advise, @Sridhar-MS @eerhardt.
Notably, I've removed vetting from the command, so NuGet can report back to the user when an unsupported command is passed, and can give usage text via the "dotnet nuget [command] --help" option. "dotnet nuget --help" already works.

brthor commented on src/dotnet/commands/dotnet-pack/NuGet/Manifest.cs:
```diff
@@ -8,7 +8,7 @@
 using System.Xml.Linq;
 
-namespace NuGet
+namespace NuGet.Legacy
```
Is this namespace change necessary? Won't we be removing this code in the near future?

drewgil replied:
I believe we will be, yes. @rohit21agrawal is undertaking the dotnet pack work (which is separate from my dotnet nuget work). This is why I've elected to use the more constistent namespace in the new code rather than adapt it to this code which will likely go away. In the meantime I need a way to resolve the conflict while waiting on the other work.

eerhardt commented:
test OSX x64 Release Build

eerhardt commented on src/dotnet/commands/dotnet-nuget/NuGetCommand.cs:
```diff
@@ -0,0 +1,73 @@
+            Debug.Assert(commandRunner != null, "A command runner must be passed to RunCommand");
+            if (commandRunner == null)
+            {
+                return -1; // invalid operation
```
(nit) I think throwing an InvalidOperation or Argument Exception is the best approach here.

eerhardt commented:
:shipit:

drewgil commented:
@Sridhar-MS, @brthor, @eerhardt  --I'm continuing to work on some other pieces here, so let's continue to discuss if there are outstanding design items for review.
